syntax = "proto3";


message PnrdPayload{
    enum Action {
        CREATE_OWNER = 0;
        CREATE_RECORD = 1;
        UPDATE_RECORD = 2;
        TRANSFER_RECORD = 3;
    }

    // Whether the payload contains a create reader, create record,
    // update record, or transfer record action
    Action action = 1;

    // The transaction handler will read from just one of these fields
    // according to the action
    CreateOwnerAction create_owner = 2;
    CreateRecordAction create_record = 3;
    UpdateRecordAction update_record = 4;
    TransferRecordAction transfer_record = 5;

    // Approximately when transaction was submitted, as a Unix UTC timestamp
    uint64 timestamp = 6;
}


message CreateOwnerAction {
    // A human-readable name identifying the new reader
    string name = 1;
}


message CreateRecordAction {
    // The user-defined natural key which identifies the object in the
    // real world (for example a serial number)
    string record_id = 1;
    string tag_id = 2;
    
    string reader_id = 3;
    string ant_id = 4;
    string situation = 5;
    int32 places = 6;
    int32 transitions = 7;
    repeated sint32 token = 8 [packed=true];
    repeated sint32 incidenceMatrix = 9 [packed=true];
}


message UpdateRecordAction {
    // The id of the record being updated
    string record_id = 1;

    string reader_id = 2;
    string ant_id = 3;
    string situation = 4;
    int32 places = 5;
    int32 transitions = 6;
    repeated sint32 token = 7 [packed=true];
    repeated sint32 incidenceMatrix = 8 [packed=true];
    
}


message TransferRecordAction {
    // The id of the record for the ownership transfer
    string record_id = 1;

    // The public key of the owner to which the record will be transferred
    string receiving_owner = 2;
}
